lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -2078,20 +2082,121 @@ void GrandCentralPass::runOnOperation() {
     VerbatimBuilder::Base verbatimData;
     VerbatimBuilder verbatim(verbatimData);
     verbatim += instanceSymbol;
-    auto iface =
-        traverseBundle(bundle, bundle.getID(), bundle.getPrefix(), verbatim);
-    if (!iface) {
+    // List of interface elements.
+
+    SmallVector<VerbatimXMRbuilder> xmrElems;
+    SmallVector<InterfaceElemsBuilder> interfaceBuilder;
+
+    auto ifaceName = traverseBundle(bundle, bundle.getID(), bundle.getPrefix(),
+                                    verbatim, xmrElems, interfaceBuilder);
+    if (!ifaceName) {
       removalError = true;
       continue;
     }
+
+    if (companionIter.isNonlocal) {
+      // If the companion module has two exactly same ViewAnnotation.companion
+      // annotations, then add the interface for only one of them. This happens
+      // when the companion is deduped.
+      auto viewMapIter = companionToInterfaceMap.find(companionModule);
+      if (viewMapIter != companionToInterfaceMap.end())
+        if (std::equal(interfaceBuilder.begin(), interfaceBuilder.end(),
+                       viewMapIter->getSecond().begin(),
+                       compareInterfaceSignal)) {
+          continue;
+        }
+
+      companionToInterfaceMap[companionModule] = interfaceBuilder;
+    }
+
+    for (auto xmrElem : xmrElems) {
+      auto builder = OpBuilder::atBlockEnd(xmrElem.companionMod.getBodyBlock());
```

Comment:
The same builder could be re-used by moving this outside the loop and using `InsertionGuard` or just being careful.

Reply:
Moved it out.